dynamiqs/utils/random.py

```diff
@@ -0,0 +1,189 @@
+        x = np.random.rand(n) * 2 - 1 + 1j * (np.random.rand(n) * 2 - 1)
+        ax0.scatter(x.real, x.imag, s=1.0)
+
+        # option 2: uniformly distributed magnitude and phase
+        x = np.random.rand(n) * np.exp(1j * 2 * np.pi * np.random.rand(n))
+        ax1.scatter(x.real, x.imag, s=1.0)
+
+        # option 3: uniformly distributed in a disk (in dynamiqs)
+        key = jax.random.PRNGKey(42)
+        x = dq.rand_complex(key, n)
+        ax2.scatter(x.real, x.imag, s=1.0)
+        renderfig('rand_complex')
+        ```
+
+        ![rand_complex](/figs-code/rand_complex.png){.fig}
+
+    Args:
+        key: A PRNG key used as the random key.
+        shape _(int or tuple of ints)_: Shape of the returned array.
+        rmax: Maximum magnitude.
+
+    Returns:
+        _(array of shape (*shape))_ Random complex number array.
+
+    Examples:
+        >>> key = jax.random.PRNGKey(42)
+        >>> dq.rand_complex(key, (2, 3), rmax=5.0)
+        Array([[ 1.341+4.17j ,  3.978-0.979j, -2.592-0.946j],
+               [-4.428+1.744j, -0.53 +1.668j,  2.582+0.65j ]], dtype=complex64)
+    """
+    shape = (shape,) if isinstance(shape, int) else shape
+    # sample uniformly in the unit L2 ball and scale
+    x = rmax * jax.random.ball(key, 2, shape=shape)
+    return x[..., 0] + 1j * x[..., 1]
+
+
+def rand_herm(key: PRNGKeyArray, shape: tuple[int, ...]) -> Array:
+    """Returns a random complex Hermitian array.
+
+    Args:
+        key: A PRNG key used as the random key.
+        shape: Shape of the returned array.
+
+    Returns:
+        _(array of shape (*shape))_ Random complex Hermitian array.
+
+    Examples:
+        >>> key = jax.random.PRNGKey(42)
+        >>> dq.rand_herm(key, (2, 2))
+        Array([[-0.291+0.j   ,  0.473-0.446j],
+               [ 0.473+0.446j,  0.13 +0.j   ]], dtype=complex64)
+    """
+    if not len(shape) >= 2 or not shape[-1] == shape[-2]:
+        raise ValueError(f'`shape` must be at least 2D and square, but got {shape}.')
+    x = rand_complex(key, shape)
+    return 0.5 * (x + dag(x))
+
+
+def rand_psd(key: PRNGKeyArray, shape: tuple[int, ...]) -> Array:
+    """Returns a random complex positive semi-definite array.
+
+    Args:
+        key: A PRNG key used as the random key.
+        shape: Shape of the returned array.
+
+    Returns:
+        _(array of shape (*shape))_ Random complex positive semi-definite array.
+
+    Examples:
+        >>> key = jax.random.PRNGKey(42)
+        >>> dq.rand_psd(key, (2, 2))
+        Array([[1.145+0.j  , 0.582+0.33j],
+               [0.582-0.33j, 0.844+0.j  ]], dtype=complex64)
+
+    """
+    if not len(shape) >= 2 or not shape[-1] == shape[-2]:
+        raise ValueError(f'`shape` must be at least 2D and square, but got {shape}.')
+    x = rand_complex(key, shape)
+    return x @ dag(x)
+
+
+def rand_dm(key: PRNGKeyArray, shape: tuple[int, ...]) -> Array:
+    """Returns a random density matrix (positive semi-definite, hermitian and unit
+    trace).
```

Comment:
```suggestion
    """Returns a random density matrix (hermitian, positive semi-definite, and unit
    trace).
```

(positive semi-definite => hermitian, so I prefer this order)